doc: fix -netinfo relaytxes help

jonatack opened the pull request:
Addresses https://github.com/bitcoin/bitcoin/pull/26109#discussion_r995502563 by Marco Falke (thanks!)

jonatack committed:
doc: fix up -netinfo relaytxes help

Co-authored-by: "MarcoFalke <*~=`'#}+{/-|&$^_@721217.xyz>"

MarcoFalke commented on src/bitcoin-cli.cpp:
```diff
@@ -642,7 +642,7 @@ class NetinfoRequestHandler : public BaseRequestHandler
         "  recv     Time since last message received from the peer, in seconds\n"
         "  txn      Time since last novel transaction received from the peer and accepted into our mempool, in minutes\n"
-        "           \"*\" - whether we relay transactions to this peer (relaytxes is false)\n"
+        "           \"*\" - we do not relay transactions to this peer (relaytxes is false or not available)\n"
```
if it is not available it will default to true, no?

shaavan replied:
@MarcoFalke 
Yes, I was able to verify your point

On line https://github.com/bitcoin/bitcoin/blob/master/src/bitcoin-cli.cpp#L542, the `"*"` or `""` is selected based on bool value of `is_tx_relay`.

```cpp
peer.last_trxn ? ToString((time_now - peer.last_trxn) / 60) : peer.is_tx_relay ? "" : "*",
```

and,
on line https://github.com/bitcoin/bitcoin/blob/master/src/bitcoin-cli.cpp#L481, `is_tx_relay` defaults to **True** when `relaytxes` is **Null**

```cpp
const bool is_tx_relay{peer["relaytxes"].isNull() ? true : peer["relaytxes"].get_bool()};
```

jonatack replied:
Thanks, done, and added displaying the new "not available" state distinctly from the "true but no last tx time" state.

jonatack commented:
@MarcoFalke @shaavan thanks for taking a look! Addressed your feedback, improved the getpeerinfo#relaytxes help, and added explicitly handling the new (v24.0) relaytxes 3rd state of "not available" in -netinfo.

shaavan reviewed (commented):
`relaytxes` is set based on the value of `m_relay_txes` if `fStateStats` is **true**; otherwise, `relaytxes` is unavailable.

[net.cpp 203]( https://github.com/bitcoin/bitcoin/blob/master/src/rpc/net.cpp#L203)

```cpp
if (fStateStats) {
            obj.pushKV("relaytxes", statestats.m_relay_txs);
        }
```

And `fStateStats` is equal to false when either Node or peer is **nullptr**:

[net_processing.cpp 1555](https://github.com/bitcoin/bitcoin/blob/master/src/net_processing.cpp#L1555)

```cpp
if (state == nullptr)
            return false;
```

[net_processing.cpp 1566](https://github.com/bitcoin/bitcoin/blob/master/src/net_processing.cpp#L1566)

```cpp
if (peer == nullptr) return false;
```

The value of `relaytxes` depends on the value of statestats.m_relay_txs.

And the value of `statestats.m_relay_txs` can be false in two cases:

if `m_tx_relay→m_relay_txs` is **false**.

[net_processing.cpp 1580](https://github.com/bitcoin/bitcoin/blob/master/src/net_processing.cpp#L1580)

```cpp
if (auto tx_relay = peer->GetTxRelay(); tx_relay != nullptr) {
        stats.m_relay_txs = WITH_LOCK(tx_relay->m_bloom_filter_mutex, return tx_relay->m_relay_txs);
```

or if `m_tx_relay` is a **nullptr**.

[net_processing.cpp 1583](https://github.com/bitcoin/bitcoin/blob/master/src/net_processing.cpp#L1583)

```cpp
else {
        stats.m_relay_txs = false;
```

And according to the comment on `m_tx_relay`, it will be a **nullptr** when we are not relaying transactions to this peer

[net_processing.cpp 398](https://github.com/bitcoin/bitcoin/blob/master/src/net_processing.cpp#L398)

```cpp
/** Transaction relay data. Will be a nullptr if we're not relaying
     *  transactions with this peer (e.g., if it's a block-relay-only peer or
     *  the peer has sent us fRelay=false with bloom filters disabled). */
    std::unique_ptr<TxRelay> m_tx_relay GUARDED_BY(m_tx_relay_mutex);
```

So as far as I understood, there are four possible states:

1. Node asked to relay, and we will relay txs to this node `(m_tx_relay` ≠ **nullptr** and `m_tx_relay` → m_relay_txs = **True**)
2. Node asked not to relay `(m_tx_relay` ≠ **nullptr** and `m_tx_relay→m_relay_txs` = **False**)
3. Node is not yet setup (`peer` == **nullptr**, irrespective of values of `m_tx_relay` and `m_relay_txs`)
4. Node is setup but we can’t relay to it (`m_tx_relay` = **nullptr**)

The first three states are properly represented by the PR using "", "*", "." repectively. However, I think there is a need to address the 4th state as well in which we can't relay transactions to the peer.

jonatack commented:
>     4. Node is setup but we can’t relay to it (`m_tx_relay` = **nullptr**)
> 
> The first three states are properly represented by the PR using "", "*", "." repectively. However, I think there is a need to address the 4th state as well in which we can't relay transactions to the peer.

@shaavan thanks for looking. Running `git grep -B4 "stats.m_relay_txs = false;"`, in the 4th case you describe `PeerManagerImpl::GetNodeStateStats()` returns `m_relay_txs` as false (and so getpeerinfo#relaytxes is false). This is one of the 3 states (true, false, not available) returned by that field and handled here.

brunoerg reviewed (commented):
utACK d51dbbc697adeac2563727e27f49b98bcc319914

I was able to verify the three states in the code and the lack of documentation that this PR fixes. Couldn't we cover this case where relaytxes is not available in a functional test? I confess that I could test manually this field by running it, and could check true/false states but it has been a little difficult to test manually and get a n/a.

luke-jr commented:
If relaytxes is omitted ONLY if it hasn't been setup yet, we should just include it as `false` IMO.

(But if it could be during peer *destruction*, that might not be correct)

jonatack commented:
@luke-jr I agree; nevertheless see https://github.com/bitcoin/bitcoin/pull/25923#issuecomment-1227233657 for the rationale, which was considered compelling enough to make getpeerinfo#relaytxes optional after being always present since the API field was introduced in 08843ed99843078acb10eecda2045d5f0f1c2b4f.

Thank you @brunoerg! @MarcoFalke @shaavan mind re-reviewing? @w0xlt @dunxen @kristapsk @1440000bytes @Zero-1729 @0xB10C @jarolrod you have reviewed these kind of changes in the past, if you don't mind having a look.

Zero-1729 commented on src/bitcoin-cli.cpp:
```diff
@@ -642,7 +643,8 @@ class NetinfoRequestHandler : public BaseRequestHandler
         "  txn      Time since last novel transaction received from the peer and accepted into our mempool, in minutes\n"
-        "           \"*\" - whether we relay transactions to this peer (relaytxes is false)\n"
+        "           \".\" - we do not relay transactions to this peer (relaytxes not available as peer connection is still being set up)\n"
+        "           \"*\" - we do not relay transactions to this peer (relaytxes is false)\n"
```
IMO, it would be worthwhile to use "`.`" to signal we aren't relaying txs (i.e. `relaytxes is false`) and have "`*`" signal the case where the info is unavailable. 

Mainly for two reasons, consistency and intuition. Regarding consistency, `addrp` already uses '`.`' to signal that we aren't relaying addresses (specifically, when `addr_relay_enabled is false`) which is a similar case here, as `relaytxes is false`. Though currently on master, `*` is used to show when `relaytxes is false`, it breaks that consistency also, and I'd say '`*`' intuitively seems more like an undetermined/pending state than '`.`'.

jonatack replied:
It's a slightly larger change of behavior, but sure -- done.

jonatack commented:
Updated to take feedback by @Zero-1729, ACKs and re-ACKs welcome :)

Zero-1729 commented on doc/release-notes-26328.md:
```diff
@@ -0,0 +1,7 @@
+Tools and Utilities
+-------------------
+
+- CLI `-netinfo` now displays 3 different states in the `txn` column (true, false, and not available), as the
```
Slight rewording for clarity

```suggestion
- CLI `-netinfo` now displays 3 different states in the `txn` column (true, false, and not available). The
```

jonatack replied:
Done

Zero-1729 commented on doc/release-notes-26328.md:
```diff
@@ -0,0 +1,7 @@
+-------------------
+
+- CLI `-netinfo` now displays 3 different states in the `txn` column (true, false, and not available), as the
+  latter value is a new, third state since 24.0 returned by the RPC getpeerinfo "relaytxes" field when the peer
```
(cont'd)

```suggestion
  last value is a new third state since 24.0 returned by the RPC getpeerinfo "relaytxes" field when the peer
```

jonatack replied:
Thanks, replaced "latter" with "not available".

Zero-1729 reviewed (commented):
tACK 7c733bd25e024a23e75fb1015e1c394df88df2d0

Tested on macOS v13.0 (22A380)

I can confirm this patch splits the displayed `relaytxes` states into three (true, false, n/a due to pending setup), updates the help docs, and release notes.

- - -

#### Brief digest diff

On master (ae6bb6e71e3082dd783e78c52b3af649fd5256cc)

```sh
$ ./src/bitcoin-cli -signet -netinfo 4   
Bitcoin Core client v24.99.0-ae6bb6e71e30-dirty signet - server 70016/Satoshi:24.99.0/

<->   type   net  mping   ping send recv  txn  blk  hb addrp addrl  age id address               version
out   full  ipv4    175    175   15   22         0       172          0  3 85.208.70.229:38333   70016/Satoshi:23.0.0/
out   full  ipv4    301    301   16   16                 123          0  4 202.61.248.219:38333  70016/Satoshi:24.99.0/
out   full  ipv4    517    517   22   22         0  .    184          0  2 72.48.253.168:38333   70016/Satoshi:22.0.0/
out   full  ipv4    570    570   10   10                 230          0  5 147.182.229.68:38333  70016/Satoshi:22.99.0/
out  block  ipv4    753    753   22   22    *    0         .          0  0 18.166.228.180:38333  70016/Satoshi:23.0.0/
out  block  ipv4    879    879   22   22    *    0  .      .          0  1 103.16.128.63:38333   70016/Satoshi:23.0.0/
out   full  ipv4   1247   1247    7    7                 205          0  6 178.128.221.177:38333 70016/Satoshi:22.99.0/
                     ms     ms  sec  sec  min  min                  min

         ipv4    ipv6   total   block
in          0       0       0
out         7       0       7       2
total       7       0       7

Local addresses: n/a
```
```sh
$ ./src/bitcoin-cli -signet getpeerinfo   
[
  {
    "id": 0,
    "addr": "18.166.228.180:38333",
    "addrbind": "192.168.0.161:53506",
    "addrlocal": "102.91.5.10:53506",
    "network": "ipv4",
    "services": "0000000000000409",
    "servicesnames": [
      "NETWORK",
      "WITNESS",
      "NETWORK_LIMITED"
    ],
    "relaytxes": false,
    …
  },
  …
  {
    "id": 2,
    "addr": "72.48.253.168:38333",
    "addrbind": "192.168.0.161:53509",
    "addrlocal": "197.210.70.183:53509",
    "network": "ipv4",
    "services": "000000000000040d",
    "servicesnames": [
      "NETWORK",
      "BLOOM",
      "WITNESS",
      "NETWORK_LIMITED"
    ],
    "relaytxes": true,
    …
  },
  …
]
```

On Patch (7c733bd25e024a23e75fb1015e1c394df88df2d0)

```sh
$ ./src/bitcoin-cli -signet -netinfo 4  
Bitcoin Core client v24.99.0-7c733bd25e02 signet - server 70016/Satoshi:24.99.0/

<->   type   net  mping   ping send recv  txn  blk  hb addrp addrl  age id address               version
out   full  ipv4    197    203   45   45                 242          4  7 49.12.208.214:38333   70016/Satoshi:0.21.1/
out   full  ipv4    198    200    2   53                 172          4  4 85.208.70.229:38333   70016/Satoshi:23.0.0/
out   full  ipv4    200    249   54   53    2            124          4  3 202.61.248.219:38333  70016/Satoshi:24.99.0/
out   full  ipv4    225    225   32   32                 148          4  9 31.14.40.19:38333     70016/Satoshi:22.0.0/
out   full  ipv4    249    284   51   51                 121          4  6 108.161.223.181:38333 70016/Satoshi:22.0.0/
out   full  ipv4    291    291   54   54    1            178          4  2 45.33.47.11:38333     70016/Satoshi:23.0.0/
out   full  ipv4    302    310   52   52                 139          4  5 104.128.228.143:38333 70016/Satoshi:23.0.0/
out   full  ipv4    303    444   44   44    3    2  .    230          4  8 147.182.229.68:38333  70016/Satoshi:22.99.0/
out  block  ipv4    387    405   62   62    .    5  .      .          5  0 18.166.228.180:38333  70016/Satoshi:23.0.0/
out  block  ipv4    442    442   61   61    .              .          5  1 103.16.128.63:38333   70016/Satoshi:23.0.0/
                     ms     ms  sec  sec  min  min                  min

         ipv4    ipv6   total   block
in          0       0       0
out        10       0      10       2
total      10       0      10

Local addresses: n/a 
```

```sh
$ ./src/bitcoin-cli -signet getpeerinfo   
[
  {
    "id": 0,
    "addr": "18.166.228.180:38333",
    "addrbind": "192.168.0.161:59682",
    "addrlocal": "197.210.71.89:64632",
    "network": "ipv4",
    "services": "0000000000000409",
    "servicesnames": [
      "NETWORK",
      "WITNESS",
      "NETWORK_LIMITED"
    ],
    "relaytxes": false,
    "lastsend": 1667626204,
    "lastrecv": 1667626204,
    "last_transaction": 0,
    …
  },
  …
  {
    "id": 4,
    "addr": "85.208.70.229:38333",
    "addrbind": "192.168.0.161:59702",
    "addrlocal": "102.91.5.10:59702",
    "network": "ipv4",
    "services": "0000000000000409",
    "servicesnames": [
      "NETWORK",
      "WITNESS",
      "NETWORK_LIMITED"
    ],
    "relaytxes": true,
    "lastsend": 1667626198,
    "lastrecv": 1667626093,
    "last_transaction": 0,
    …
  },
  …
 {
    "id": 8,
    "addr": "147.182.229.68:38333",
    "addrbind": "192.168.0.161:59717",
    "addrlocal": "197.210.71.89:59717",
    "network": "ipv4",
    "services": "0000000000000409",
    "servicesnames": [
      "NETWORK",
      "WITNESS",
      "NETWORK_LIMITED"
    ],
    "relaytxes": true,
    "lastsend": 1667626197,
    "lastrecv": 1667626102,
    "last_transaction": 1667625918,
    …
  }
  …
]
```

jonatack commented:
Updated to take feedback by @Zero-1729, ACKs and re-ACKs welcome :)

Zero-1729 reviewed (approved):
crACK cfabd15309eb42a9f20f4c11152c54fd27978fab

mzumsande commented on src/bitcoin-cli.cpp:
```diff
@@ -642,7 +643,8 @@ class NetinfoRequestHandler : public BaseRequestHandler
         "  txn      Time since last novel transaction received from the peer and accepted into our mempool, in minutes\n"
-        "           \"*\" - we do not relay transactions to this peer (relaytxes is false)\n"
+        "           \".\" - we do not relay transactions to this peer (relaytxes is false)\n"
+        "           \"*\" - we do not relay transactions to this peer (relaytxes not available as peer connection is still being set up)\n"
```
See https://github.com/bitcoin/bitcoin/pull/26457#discussion_r1024158792, I think the third state (N/A) would not be used during connection startup as intended. I confirmed this by adding a sleep at the beginning of `::VERSION `processing in `ProcessMessage` and querying getpeerinfo with this branch - I'd get `relaytxes= false` but the field is included in the response.

jonatack commented:
Closing for now in favor of #26516.

jonatack commented:
Re-opening after reading https://github.com/bitcoin/bitcoin/pull/26457#issuecomment-1316597000, pending reviewer feedback here and on #26516 (the latter has the benefit of not changing user space).

MarcoFalke commented:
I think how RPC returns that a value is missing, (with optional, or a magic value, or some extended logic) is orthogonal to cli netinfo changes. However, it would be beneficial to first figure out the RPC interface and then adjust the cli and gui code to that.

jonatack commented:
Simplified to just the fixup to https://github.com/bitcoin/bitcoin/pull/26109#discussion_r995502563 found by @MarcoFalke.

mzumsande reviewed (commented):
Code Review ACK 0f5fc4f656d0990802ab552c0e620f49e785fed4

MarcoFalke commented:
I think my misunderstanding came from the fact that it is unclear if `version.relaytxs` or `getpeerinfo.relaytxs` was meant, see https://github.com/bitcoin/bitcoin/pull/26109/files#r1067201660

In any case https://github.com/bitcoin/bitcoin/commit/0f5fc4f656d0990802ab552c0e620f49e785fed4 lgtm

fanquake commented:
This has been merged.

jonatack commented:
> I think my misunderstanding came from the fact that it is unclear if `version.relaytxs` or `getpeerinfo.relaytxs` was meant, see [#26109 (files)](https://github.com/bitcoin/bitcoin/pull/26109/files#r1067201660)
> 
> In any case [0f5fc4f](https://github.com/bitcoin/bitcoin/commit/0f5fc4f656d0990802ab552c0e620f49e785fed4) lgtm

I almost added `s/relaytxes/getpeerinfo \"relaytxes\" field/` here yesterday. Might sneak it next time I touch the file.
